lib/executor/ramping_vus.go

```diff
@@ -636,40 +633,88 @@ func (vlv RampingVUs) Run(ctx context.Context, out chan<- stats.SampleContainer)
 		currentMaxAllowedVUs = newMaxAllowedVUs
 	}
 
-	handleAllRawSteps := func() bool {
-		for {
+	// This is the step that is the actual gracefulStop at the end
+	lastStep := gracefulExecutionSteps[len(gracefulExecutionSteps)-1]
+	// TODO maybe instead of getting only the last step (above) mergeSteps should return the
+	// remaining gracefulExecutionSteps? or maybe we should refactor this even more ...
+	mergedSteps := mergeSteps(rawExecutionSteps, gracefulExecutionSteps[:len(gracefulExecutionSteps)-1])
+	timer := time.NewTimer(time.Hour * 24)
+	for _, step := range mergedSteps {
+		offsetDiff := step.step.TimeOffset - time.Since(startTime)
+		if offsetDiff > 0 { // wait until time of event arrives // TODO have a mininum
+			timer.Reset(offsetDiff)
 			select {
-			case step, ok := <-rawStepEvents:
-				if !ok {
-					return true
-				}
-				handleNewScheduledVUs(step.PlannedVUs)
-			case step := <-gracefulLimitEvents:
-				if step.PlannedVUs > currentMaxAllowedVUs {
-					// Handle the case where a value is read from the
-					// gracefulLimitEvents channel before rawStepEvents
-					handleNewScheduledVUs(step.PlannedVUs)
-				}
-				handleNewMaxAllowedVUs(step.PlannedVUs)
-			case <-ctx.Done():
-				return false
+			case <-maxDurationCtx.Done():
+				return // exit if context is cancelled
+			case <-timer.C:
+				// now we do a step
 			}
 		}
+		if step.raw {
+			handleNewScheduledVUs(step.step.PlannedVUs)
+		} else {
+			handleNewMaxAllowedVUs(step.step.PlannedVUs)
+		}
 	}
 
-	if handleAllRawSteps() {
-		// Handle any remaining graceful stops
-		go func() {
-			for {
-				select {
-				case step := <-gracefulLimitEvents:
-					handleNewMaxAllowedVUs(step.PlannedVUs)
-				case <-maxDurationCtx.Done():
-					return
-				}
+	go func() {
+		offsetDiff := lastStep.TimeOffset - time.Since(startTime)
+		if offsetDiff > 0 { // wait until time of event arrives // TODO have a mininum
+			timer.Reset(offsetDiff)
+			select {
+			case <-maxDurationCtx.Done():
+				return // exit if context is cancelled
+			case <-timer.C:
+				// now we do a step
 			}
-		}()
-	}
+		}
+		handleNewMaxAllowedVUs(lastStep.PlannedVUs)
+	}()
 
 	return nil
 }
+
+type mergedStep struct {
+	raw  bool              // marks if this is a raw step or gracefulStop one
+	step lib.ExecutionStep // the actual step
+}
+
+// merges raw and gracefulStop steps, giving priority to raw ones if time offset is the same
+// TODO ... thinking about it gracefulExecutionSteps in a lot of cases will lead to nothing so maybe
```

Comment:
Yeah, the current code would save a lot of the ramp-up steps twice, since they are present in both sets of steps. Not a big deal, and can be addressed at the same time as https://github.com/loadimpact/k6/issues/1499

Reply:
yeah, I don't want to go rewrite the whole generation code ... IMO there should ramp-up steps and very few gracefulStop ones ... but I might be wrong, and I don't want to go down the rabbit hole of rewriting it right now.

Reply:
fixed in 51b969a